base/dcp_dest.go

```diff
@@ -0,0 +1,406 @@
+package base
+
+import (
+	"context"
+	"errors"
+	"expvar"
+	"fmt"
+	"io"
+	"strconv"
+
+	"github.com/couchbase/cbgt"
+	"github.com/couchbase/gomemcached"
+	sgbucket "github.com/couchbase/sg-bucket"
+	pkgerrors "github.com/pkg/errors"
+)
+
+// vbucketIdStrings is a memorized array of 1024 entries for fast
+// conversion of vbucketId's to partition strings via an index lookup.
+// (Atoi is faster than map lookup when going in the other direction)
+var vbucketIdStrings []string
+
+func init() {
+	vbucketIdStrings = make([]string, 1024)
+	for i := 0; i < len(vbucketIdStrings); i++ {
+		vbucketIdStrings[i] = fmt.Sprintf("%d", i)
+	}
+}
+
+type SGDest interface {
+	cbgt.Dest
+	initFeed(backfillType uint64) error
+}
+
+// DCPDest implements SGDest (superset of cbgt.Dest) interface to manage updates coming from a
+// cbgt-based DCP feed.  Embeds DCPCommon for underlying feed event processing
+type DCPDest struct {
+	DCPCommon
+}
+
+func NewDCPDest(callback sgbucket.FeedEventCallbackFunc, bucket Bucket, maxVbNo uint16, persistCheckpoints bool, dbStats *expvar.Map, feedID string) (SGDest, context.Context) {
+
+	dcpCommon := NewDCPCommon(callback, bucket, maxVbNo, persistCheckpoints, dbStats, feedID)
+
+	d := &DCPDest{
+		DCPCommon: dcpCommon,
+	}
+
+	if LogDebugEnabled(KeyDCP) {
+		InfofCtx(d.loggingCtx, KeyDCP, "Using DCP Logging Receiver")
+		logRec := &DCPLoggingDest{dest: d}
+		return logRec, d.loggingCtx
+	}
+
+	return d, d.loggingCtx
+}
+
+func (d *DCPDest) Close() error {
+	// TODO: Do we have any cleanup we should be doing on close?
+	return nil
+}
+
+func (d *DCPDest) DataUpdate(partition string, key []byte, seq uint64,
+	val []byte, cas uint64, extrasType cbgt.DestExtrasType, extras []byte) error {
+
+	if !dcpKeyFilter(key) {
+		return nil
+	}
+	event := makeFeedEventForDest(key, val, cas, partitionToVbNo(partition), 0, 0, sgbucket.FeedOpMutation)
+	d.dataUpdate(seq, event)
+	return nil
+}
+
+func (d *DCPDest) DataUpdateEx(partition string, key []byte, seq uint64, val []byte,
+	cas uint64, extrasType cbgt.DestExtrasType, req interface{}) error {
+
+	if !dcpKeyFilter(key) {
+		return nil
+	}
+
+	var event sgbucket.FeedEvent
+	if extrasType == cbgt.DEST_EXTRAS_TYPE_MCREQUEST {
+		mcReq, ok := req.(*gomemcached.MCRequest)
+		if !ok {
+		}
```

Comment:
Should probably be returning an error here?

```suggestion
 			return errors.New("Unable to cast extras of type DEST_EXTRAS_TYPE_MCREQUEST to gomemcached.MCRequest")
		}
```